// ytdl-core-new/dist/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const stream_1 = require("stream");
const info = __importStar(require("./info"));
const util = __importStar(require("./util"));
const sig = __importStar(require("./sig"));
const axios_1 = __importDefault(require("axios"));
const m3u8stream_1 = __importDefault(require("m3u8stream"));
const parse_time_1 = __importDefault(require("m3u8stream/dist/parse-time"));
const client_1 = __importDefault(require("./client"));
const ytdl = (link, options) => {
    const stream = createStream(options);
    exports.getInfo(link, options)
        .then((info) => {
        stream.emit('info', info);
        return downloadFromInfoCallback(stream, info, options);
    })
        .catch((e) => stream.emit('error', e));
    return stream;
};
exports.ytdl = ytdl;
const createStream = (options) => {
    var _a, _b;
    const stream = new stream_1.PassThrough({
        highWaterMark: (_b = (_a = options) === null || _a === void 0 ? void 0 : _a.highWaterMark, (_b !== null && _b !== void 0 ? _b : null)),
    });
    stream.destroy = () => {
        stream._isDestroyed = true;
    };
    return stream;
};
/**
 * Chooses a format to download.
 *
 * @param {stream.Readable} stream
 * @param {Object} info
 * @param {Object} options
 */
const downloadFromInfoCallback = (stream, info, options) => __awaiter(void 0, void 0, void 0, function* () {
    options = options || {};
    const format = util.chooseFormat(info.formats, options);
    if (format instanceof Error) {
        // The caller expects this function to be async.
        setImmediate(() => {
            stream.emit('error', format);
        });
        return;
    }
    stream.emit('info', { info, format });
    if (stream._isDestroyed) {
        return;
    }
    let contentLength = 0;
    let downloaded = 0;
    const ondata = (chunk) => {
        downloaded += chunk.length;
        stream.emit('progress', {
            chunkLength: chunk.length,
            downloaded,
            total: contentLength,
        });
    };
    let reqType;
    let req;
    if (format.isHLS || format.isDashMPD) {
        reqType = 'miniget';
        req = m3u8stream_1.default(format.url, {
            chunkReadahead: +info.live_chunk_readahead,
            begin: (options.begin || format.live && Date.now()),
            liveBuffer: options.liveBuffer,
            requestOptions: (options.requestOptions),
            parser: format.isDashMPD ? 'dash-mpd' : 'm3u8',
            id: format.itag,
        });
        req.on('progress', (segment, totalSegments) => {
            stream.emit('progress', {
                chunkLength: segment.size,
                downloaded: segment.num,
                total: totalSegments,
            });
        });
    }
    else {
        if (options.begin) {
            format.url += '&begin=' + parse_time_1.default.humanStr(options.begin);
        }
        let requestOptions = Object.assign({}, options.requestOptions);
        if (options.range && (options.range.start || options.range.end)) {
            requestOptions.headers = Object.assign({}, requestOptions.headers, {
                Range: `bytes=${options.range.start || '0'}-${options.range.end || ''}`,
            });
        }
        reqType = 'axios';
        const source = axios_1.default.CancelToken.source();
        req = yield client_1.default.get(format.url, Object.assign(Object.assign({}, requestOptions), { headers: Object.assign({ 'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.106 Safari/537.36', Referer: `https://www.youtube.com/watch/${info.video_id}` }, requestOptions.headers), responseType: 'stream', cancelToken: source.token }));
        stream.emit('response', req);
        req.abort = () => source.cancel();
        if (stream._isDestroyed) {
            return;
        }
        if (!contentLength) {
            contentLength = Number(req.headers['content-length']);
        }
        req.data.on('data', ondata);
    }
    stream.destroy = () => {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        stream._isDestroyed = true;
        if (req.abort)
            req.abort();
        (_b = (_a = req).end) === null || _b === void 0 ? void 0 : _b.call(_a);
        (_d = (_c = req).removeListener) === null || _d === void 0 ? void 0 : _d.call(_c, 'data', ondata);
        if (reqType === 'miniget') {
            (_f = (_e = req).unpipe) === null || _f === void 0 ? void 0 : _f.call(_e);
        }
        else {
            (_h = (_g = req.data).unpipe) === null || _h === void 0 ? void 0 : _h.call(_g);
        }
    };
    if (reqType === 'miniget') {
        // Forward events from the request to the stream.
        [
            'abort', 'request', 'response', 'error', 'retry', 'reconnect',
        ].forEach((event) => {
            var _a, _b;
            (_b = (_a = req).prependListener) === null || _b === void 0 ? void 0 : _b.call(_a, event, (arg) => {
                stream.emit(event, arg);
            });
        });
        req.pipe(stream);
    }
    else {
        req.data.pipe(stream);
    }
});
exports.downloadFromInfo = (info, options) => {
    const stream = createStream(options);
    if (!info.full) {
        throw new Error('Cannot use `ytdl.downloadFromInfo()` when called ' +
            'with info from `ytdl.getBasicInfo()`');
    }
    downloadFromInfoCallback(stream, info, options);
    return stream;
};
exports.default = ytdl;
exports.getBasicInfo = info.getBasicInfo;
exports.getInfo = info.getFullInfo;
exports.chooseFormat = util.chooseFormat;
exports.filterFormats = util.filterFormats;
exports.validateID = util.validateID;
exports.validateURL = util.validateURL;
exports.getURLVideoID = util.getURLVideoID;
exports.getVideoID = util.getVideoID;
exports.cache = Object.freeze({
    sig: sig.cache,
    info: info.cache,
});
